repository_service_tuf_worker/repository.py

```diff
@@ -481,6 +485,98 @@ def _send_publish_targets_task(
             acks_late=True,
         )
 
+    def _bootstrap_online_roles(self, root: Metadata[Root], task_id: str):
+        """
+        Bootstrap the roles that uses the online key
+        """
+        _keyid: str = root.signed.roles[Timestamp.type].keyids[0]
+        public_key = root.signed.keys[_keyid]
+
+        # Top level roles (`Targets`, `Timestamp``, `Snapshot`) initialization
+        targets: Metadata[Targets] = Metadata(Targets())
+        snapshot: Metadata[Snapshot] = Metadata(Snapshot())
+        timestamp: Metadata[Timestamp] = Metadata(Timestamp())
+
+        # Calculate the bit length (Number of bits between 1 and 32)
+        bit_length = int(
+            log(self._settings.get_fresh("NUMBER_OF_DELEGATED_BINS"), 2)
+        )
+        # Succinct delegated roles (`bins`)
+        succinct_roles = SuccinctRoles([], 1, bit_length, BINS)
+        targets.signed.delegations = Delegations(
+            keys={}, succinct_roles=succinct_roles
+        )
+        # Initialize all succinct delegated roles (`bins`), update expire,
+        # sign, add to `Snapshot` meta and persist in the backend storage
+        # service.
+        for delegated_name in succinct_roles.get_roles():
+            targets.signed.add_key(
+                SSlibKey.from_securesystemslib_key(
+                    self._key_storage_backend.get(public_key).key_dict
+                ),
+                delegated_name,
+            )
+            bins_role = Metadata(Targets())
+            self._bump_expiry(bins_role, BINS)
+            self._sign(bins_role)
+            snapshot.signed.meta[f"{delegated_name}.json"] = MetaFile(
+                version=bins_role.signed.version
+            )
+            self._persist(bins_role, delegated_name)
+
+        # Create all Target Roles in the database RSTUFTargetRoles
+        db_target_roles = [
+            targets_schema.RSTUFTargetRoleCreate(
+                rolename=target_role, version=1
+            )
+            for target_role in succinct_roles.get_roles()
+        ]
+        targets_crud.create_roles(self._db, db_target_roles)
+
+        # Update `Snapshot` meta with targets
+        snapshot.signed.meta[f"{Targets.type}.json"] = MetaFile(
+            version=targets.signed.version
+        )
+
+        # Update expire, sign and persist the top level roles (`Targets`,
+        # `Timestamp``, `Snapshot`) in the backend storage service.
+        for role in [targets, snapshot, timestamp]:
+            self._bump_expiry(role, role.signed.type)
+            self._sign(role)
+            self._persist(role, role.signed.type)
+
+        self.write_repository_settings("ROOT_SIGNING", None)
+        self.write_repository_settings("BOOTSTRAP", task_id)
+        logging.info(f"Bootstrap finished with id {task_id}")
+
+    def _validate_signatures(self, metadata: Metadata, rolename: str) -> bool:
```

Comment:
I'm unsure if we can use the `Metadata.verify_delegate`. 
Using it, I need to handle all the exceptions `UnsignedMetadataError`, `KeyError`, `ValueError` to return a clear status to the user.
- At least one signature is required in the Metadata
- The signature can be not an authorized one
- The signature keyid is not a key for signature

I'm open to handling it better, maybe we can sync.

Reply:
I just took a closer look at `_validate_signatures` and I'm convinced that you should be using `Metadata.verify_delegate`.   Handling the exceptions shouldn't be more (complex) code, than what you currently have there.

Besides, the function you have currently only verifies self-signed metadata. This works for bootstrap (root v1), but won't work for subsequent metadata update, neither for root nor for targets. Because root also requires keys and threshold info from previous root. And targets requires keys and threshold only from root (or a delegating targets role).